### PR TITLE
Improvements in Theta Sketch to (hopefully) eliminate the

### PR DESCRIPTION
**Pull request by leerho**
Misinterpretation of SingleItemSketches. 

This was driven by Hive issue Number 50.
We should try to test this in the field.

### CODE REVIEW

**Comment by leerho** [comment]
Yes, the SingleItemSketch is never directly created by the user. It is a type of CompactSketch and results from a getResults() from a set operation, union or by a user compacting one of the updatable Theta Sketches.  Before the SIS, a normal CompactSketch with a single item consumed 24 bytes.  The SIS consumes only 16 bytes, which is a reduction of 33%.  Considering that the distribution of sketch sizes is predominantly power-law (long-tail) means if you have a millions of sketches the vast majority will have just one item.  So the reduction in storage for all of these single item sketches can be significant.

File: src/main/java/org/apache/datasketches/theta/Sketch.java [hunk 1]
```diff
@@ -677,15 +677,14 @@ private static final Sketch heapifyFromMemory(final Memory srcMem, final long se
           throw new SketchesArgumentException(
               "Corrupted: COMPACT family sketch image must have Read-Only flag set");
         }
-        final boolean empty = PreambleUtil.isEmpty(srcMem); //also checks memCap < 16
-        if (empty) { //empty flag or < 16 bytes
+        final boolean empty = PreambleUtil.isEmpty(srcMem);
+        if (empty) { //TODO
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/theta/Sketch.java:681]: not clear what to do

File: src/main/java/org/apache/datasketches/theta/UnionImpl.java [hunk 2]
```diff
@@ -431,6 +426,10 @@ private void processVer3(final Memory skMem) {
   //has seedHash and p, could have 0 entries & theta,
   // can only be compact, ordered, size >= 8
   private void processVer2(final Memory skMem) {
+    final int famId = extractFamilyID(skMem);
+    if (famId != 3) {
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/theta/UnionImpl.java:430]: should this refer to a constant in Family?

File: src/test/java/org/apache/datasketches/theta/EmptyTest.java [hunk 3]
```diff
@@ -111,45 +105,6 @@ public void checkPsampling() {
     assertEquals(sk1.compact().toByteArray().length, 8);
   }
 
-  //These 3 tests reproduce a failure mode where an "old" empty sketch of 8 bytes without
```
Comment (AlexanderSaydakov) [inline src/test/java/org/apache/datasketches/theta/EmptyTest.java:114]: why these tests are removed?

**Comment by leerho** [comment]
That's funny, I thought I replied to each of your comments, but they didn't show up.  Maybe I forgot to hit "Comment" ? :)

File: src/test/java/org/apache/datasketches/theta/EmptyTest.java (hunk 3 above)
Comment (leerho) [reply src/test/java/org/apache/datasketches/theta/EmptyTest.java]: This was an accident.  Replaced them.

File: src/main/java/org/apache/datasketches/theta/UnionImpl.java (hunk 2 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/theta/UnionImpl.java]: Fixed.

File: src/main/java/org/apache/datasketches/theta/Sketch.java (hunk 1 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/theta/Sketch.java]: This has been fixed.

File: src/main/java/org/apache/datasketches/theta/PreambleUtil.java [hunk 4]
```diff
@@ -462,6 +468,20 @@ static boolean isEmpty(final Memory mem) {
...
+  static boolean isSingleItem(final Memory mem) {
+    // Flags byte must be LittleEndian, ReadOnly, Not Empty, Compact, Ordered = 11010 = 0x1A.
+    // Flags mask will be 0x1F.
+    // SingleItem flag may not be set due to a historical bug, so we can't depend on it for now.
+    // However, if the above flags are correct, preLongs == 1, SerVer >= 3, FamilyID == 3,
+    // and the hash seed matches (not done here), it is virtually guaranteed that we have a
+    // SingleItem Sketch.
+    final boolean preLongs = extractPreLongs(mem) == 1;
+    final boolean serVer = extractSerVer(mem) >= 3;
+    final boolean famId = extractFamilyID(mem) == 3; //compact
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/theta/PreambleUtil.java]: Fixed.

File: src/main/java/org/apache/datasketches/theta/HeapAnotB.java [hunk 5]
```diff
@@ -99,10 +99,9 @@ public CompactSketch getResult(final boolean dstOrdered, final WritableMemory ds
       Arrays.sort(compactCache);
     }
     //Create the CompactSketch
-    final boolean empty = (curCount_ == 0) && (thetaLong_ == Long.MAX_VALUE);
     final CompactSketch comp = createCompactSketch(
-        compactCache, empty, seedHash_, curCount_, thetaLong_, dstOrdered, dstMem);
-    reset();
+        compactCache, empty_, seedHash_, curCount_, thetaLong_, dstOrdered, dstMem);
+    reset(); //TODO
```
Comment (AlexanderSaydakov) [inline src/main/java/org/apache/datasketches/theta/HeapAnotB.java:104]: not clear what to do here

Comment (leerho) [reply src/main/java/org/apache/datasketches/theta/HeapAnotB.java:104]: That is a placeholder for the next round of changes.  Just ignore.
Thanks for the review!

**Pull request merged**